Embed Block: Border support added

benazeer-ben opened the pull request:
## What?

Adding border support for Embed block.
Part of:  https://github.com/WordPress/gutenberg/issues/43247

## Why?

Embed block is missing border support


## How?

Adds the border support via block.json


## Testing Instructions

- Go to the Global Styles setting.
- Make sure that the Embed block's border is configurable via Global Styles.
- Verify that Global Styles are applied correctly in the editor and frontend.
- Navigate to post/page edit.
- Then add Embed block and apply the border styles.
- Verify that block border styles take precedence over global block colour settings.
- Verify that block borders display correctly in both the editor and frontend.
- Border will get applied only to the embeds, not to captions.

NOTE:

1. WP embeds are handling in bit different way, so make sure all embeds are working well.  You can try with all possible embeds available like YouTube, X(Twitter), Spotify, SoundCloud and WordPress posts.
2. This PR does not handle border-radius since it requires padding support as well. Adding padding support would involve reconsidering several scenarios, which could complicate the implementation. To keep this PR straightforward, border-radius and padding are not addressed at this time.


### Sample URLs for Testing:

WordPress Post:
https://wordpress.org/documentation/article/embeds/

YouTube Video:
https://www.youtube.com/watch?v=sYNqeahE2rg

Tweet:
https://x.com/WordPress

Spotify Playlist:
https://open.spotify.com/playlist/37i9dQZF1DXcBWIGoYBM5M

SoundCloud Track:
https://soundcloud.com/forss/flickermood

## Screenshots or screencast 


https://github.com/user-attachments/assets/53186fed-812b-4fe4-9b14-57a45745d3e4

benazeerhassan1909 committed:
Border support added for Embed block

benazeerhassan1909 committed:
Border support added for Embed block

benazeerhassan1909 committed:
Merge branch 'add/embed-border' of https://github.com/benazeer-ben/gutenberg into add/embed-border

benazeer-ben committed:
Revert "Style Book: Add color tab (#65692)"

This reverts commit 54371146e6515a15118c6feceb34ade061b44a1b.

benazeer-ben committed:
Merge branch 'trunk' into add/embed-border

akasunil committed:
Synced with trunk and resolved conflicts

akasunil committed:
Revert unneccessory changes

aaronrobertshaw reviewed (commented):
Thank you for putting this together @benazeer-ben 🙇 

I get the feeling this one is going to have to wrangle several edge cases given the various types of content the block can embed e.g. youtube videos, tweets, spotify and audio links etc.

Testing this PR as is, I've run into a couple of issues:

1. To support border radius we should probably also support padding on the block. This might need design feedback as discussed on [Video: Add border support](https://github.com/WordPress/gutenberg/pull/63777#pullrequestreview-2232172335).
2. The current approach in this PR applies the border to the block's wrapper. The embed block can also include a caption. Other blocks such as the image block, apply the border to the actual element that needs the border leaving the caption outside it. I think this would be the expected behaviour here as well.
3. When adding border or padding support we need to ensure the block has `box-sizing: border-box` styles so that it will respect any width its parent might be trying to apply to it.

#### Box sizing issue:

<img width="400" alt="Screenshot 2024-11-11 at 4 19 09 pm" src="https://github.com/user-attachments/assets/80e2e602-ecab-43ef-bad5-37548631fac8">

#### Embed blocks with different content and captions:


<img width="400" alt="Screenshot 2024-11-11 at 4 30 31 pm" src="https://github.com/user-attachments/assets/df884789-3856-4cf8-a5b6-1fa62e6dcfdc">

#### Next Steps

To move this PR forward while waiting on some design input regarding border-radius adoption, I think we can:

1. Skip serialization of borders on the block wrapper and apply it to the embed's inner wrapper so the caption falls outside it. See https://github.com/WordPress/gutenberg/pull/63777 for some clues on this approach.
2. Add `box-sizing: border-box` to the block. (there's also a separate PR aimed at generically adding this style to blocks but it might be slow to land given the potential to cause visual regressions).

Thanks again for all your work on adopting block supports, it's appreciated 👍

benazeer-ben committed:
Feedback changes - Skip serialization & border box

benazeer-ben commented:
Thanks @aaronrobertshaw for the detailed feedback and details.

Implemented skip serialization of borders on the block wrapper and applied it to the embed's inner wrapper.

For now padding support also added, if we get some design feedback we can change it accordingly.

Please let me know if we need more tweak on this.


https://github.com/user-attachments/assets/45146da5-01c2-4e13-a026-7285a6121137

aaronrobertshaw commented on packages/block-library/src/embed/block.json:
```diff
@@ -43,12 +43,29 @@
 		}
 	},
+	"selectors": {
+		"border": ".wp-block-embed__wrapper "
```
The convention with these selectors is to explicitly include the block's class. In this case, `.wp-block-embed`. 

I'm not 100% whether it is required but it does make it clearer the intended target for the styles. That block class is also tweaked when generating block style variation selectors for the block. 

One last nit, there's an unnecessary space at the end of that selector. Better safe than sorry, we should clean that up in case theres some theme.json processing somewhere that wasn't expecting it.

aaronrobertshaw commented on packages/block-library/src/embed/block.json:
```diff
@@ -43,12 +43,29 @@
 		"spacing": {
-			"margin": true
+			"margin": true,
+			"padding": true
```
This applies the padding to the Embed block's wrapper. 

The justification for adding the padding support was so that it could be used to counter border radius and prevent the content from sticking out beyond the border.

This might need to be applied to the inner embed wrapper as well.

aaronrobertshaw reviewed (commented):
Thank you for iterating on this 👍 

I've given it a test and there are still a few issues. Most notably:

1. Padding is applied to the wrapper so doesn't help with the border radius problem
2. The custom selector for borders should probably also have the block class in it to ensure compatibility with block style variations. 
3. PR description and test instructions no longer match what the PR does

On that last point, I find it handy each time I iterate on a PR to re-test it in full asking myself some questions like:
- Have I changed what the PR does?
- If it does something new, how do I test that?
- If the PR is related to a block, are there different ways to use the block? e.g. testing the embed block with not just a video but all the [other embed types](https://wordpress.org/documentation/article/embed-block/).

Answering those questions then means you already have everything you need to update the PR description in a way that ensures things don't get missed and it's easier for reviewers.

--- 

Here's what I see when applying both padding and borders to various embeds. I'd expect the padding within the border, not outside it otherwise it's basically just a second margin.

<img width="1289" alt="Screenshot 2024-11-18 at 6 37 34 pm" src="https://github.com/user-attachments/assets/6825d8e9-a250-469b-b905-f3e5fa5b0888">

benazeer-ben commented:
Hi @aaronrobertshaw 

Thank you for your detailed feedback and for testing the PR! 🙏

To keep things straightforward and manageable, how about we focus only on the border support for this PR and hold off on padding and border radius for now? This approach should help us avoid any complications and keep the scope of this PR more focused.

Apologies for the confusion caused by the current PR description and test instructions—I’ll make a note to clarify these aspects in future PRs. Your suggestions on updating the PR description through iterative testing are very helpful, and I'll incorporate this approach going forward to ensure clarity and completeness.

Thank you again for your patience and insights!

aaronrobertshaw commented:
> To keep things straightforward and manageable, how about we focus only on the border support for this PR and hold off on padding and border radius for now? 

As the initial intent was to adopt border support including border-radius (which I think is expected by users), padding is a natural inclusion here. I think there is some benefit to introducing both at the same time where testing can cover the whole picture and how the supports interact together.

> This approach should help us avoid any complications and keep the scope of this PR more focused.

In some ways I agree, however, there are some curly problems around deprecations and block supports, especially if the application of block supports might require skipping serialization. To me, that risk means it's better to work through the whole process here holistically.

Additionally, you've already gone to the efforts of skipping serialization of border styles to apply them to the inner embed wrapper. It's not much more to do the same for padding support. The `__experimentalSkipSerialization` supports flag also supports arrays so that individual features can be skipped while the rest are still applied to the wrapper (in this case, margin would still go on the outer wrapper).

Some examples of individual features skipping serialization are:
- Cover gradients
- Navigation text-decoration
- Gallery block gap
- Calendar text & background colors

Lastly, adding borders to the Embed block is a bit of a niche use case. This means there isn't a pressing timeline to deliver the feature. We have the luxury to take our time, dot our i's and cross out t's.

Are you onboard with that?

benazeer-ben commented:
@aaronrobertshaw Thank you for your thoughtful response and for outlining the considerations so clearly. 

I agree that addressing border and padding together provides a more complete solution, especially given how they interact. The examples of __experimentalSkipSerialization for individual features are very helpful and will guide the implementation.

I'll proceed with implementing the feature, ensuring deprecations and block supports are handled properly while keeping the scope clear and test instructions comprehensive. 

Thanks again for guidance— I’ll keep updated as I progress!

benazeer-ben committed:
Skip serialization for padding

aaronrobertshaw reviewed (commented):
Thanks for highlighting that issue and the detailed info @benazeer-ben 👍 

This is a tricky one!

There could be a few options to explore to allow supporting padding e.g. an extra wrapper, using flex/grid positioning, transforms for the iframe. All of those options have their own pros and cons whether that is backward compatibility, needing deprecations, or impacting the aspect ratio support. This would be better suited in a follow-up PR.

Given the complexity uncovered here I think your earlier suggestion to avoid padding support and focus on border styles is a good one. We'll need to omit support for `border-radius` also.

Border styles will still need to be applied to the inner wrapper and I don't think that will change. In a follow-up that pursues border radius support, I'd expect that to also be applied to the same element, so we shouldn't be painting ourselves into a corner here.

Let's go with the plan to omit padding and border-radius support so we can at least offer some border support in the short term.

How's that sound to you?

benazeer-ben commented:
Thank you for the detailed breakdown! 

I agree with your assessment—introducing padding support, especially given the iframe's positioning complexities, could indeed lead to challenges with backward compatibility and aspect ratio handling. Focusing on border styles while omitting padding and border-radius support for now sounds like a practical approach to deliver incremental improvements without overcomplicating this PR.

I’m on board with this plan and will proceed accordingly, ensuring border styles are applied to the inner wrapper as discussed. For border-radius and padding, I agree these can be revisited in a follow-up PR with a more comprehensive approach.

Thanks again for the clear direction! Let me know if there’s anything else to adjust before moving forward.

benazeer-ben commented:
Thank you for the detailed breakdown! I agree with your assessment—introducing padding support, especially given the iframe's positioning complexities, could indeed lead to challenges with backward compatibility and aspect ratio handling. Focusing on border styles while omitting padding and border-radius support for now sounds like a practical approach to deliver incremental improvements without overcomplicating this PR.

I’m on board with this plan and will proceed accordingly, ensuring border styles are applied to the inner wrapper as discussed. For border-radius and padding, I agree these can be revisited in a follow-up PR with a more comprehensive approach.

Thanks again for the clear direction!

benazeer-ben committed:
Doc Build

benazeer-ben committed:
Update for wordpress embed case border styling

aaronrobertshaw commented on packages/block-library/src/embed/block.json:
```diff
@@ -47,8 +47,24 @@
 			"clientNavigation": true
+		},
+		"__experimentalBorder": {
+			"radius": false,
```
The border support is opt-in by default so we shouldn't have to explicitly set it to `false`. That said, this does communicate that it was a conscious decision to omit border-radius support so we can leave it..

aaronrobertshaw commented on packages/block-library/src/embed/block.json:
```diff
@@ -47,8 +47,24 @@
+			"width": true,
+			"style": true,
+			"__experimentalDefaultControls": {
+				"radius": false,
```
```suggestion
```

The fact that border-radius support is explicitly disabled is enough.

aaronrobertshaw commented on packages/block-library/src/embed/embed-preview.js:
```diff
@@ -64,16 +68,27 @@ export default function EmbedPreview( {
-			<WpEmbedPreview html={ html } />
+			<WpEmbedPreview
+				html={ html }
+				className={ borderProps.className }
+				style={ { ...borderProps.style } }
```
I'm not sure about these new props. Would it be more forward-compatible for them to be combined into something like `wrapperProps`? It might also help communicate the intended use for the class name and style.

It might also help get other reviewers if we can update the test instructions to cover this scenario as well as others (e.g. explaining to test with a WordPress post as an embed, youtube video, tweet, spotify link etc.) Even better if you could include some test URLs 🙏

benazeer-ben replied:
Done in last commit.

aaronrobertshaw commented on packages/block-library/src/embed/wp-embed-preview.js:
```diff
@@ -13,7 +13,7 @@ const attributeMap = {
 };
 
-export default function WpEmbedPreview( { html } ) {
+export default function WpEmbedPreview( { html, className, style } ) {
```
As noted in my earlier comment, these new props might be better combined into a single `wrapperProps` prop or something.

benazeer-ben replied:
Done in last commit.

aaronrobertshaw reviewed (commented):
We're slowly narrowing in on something we can land. Thanks for your patience @benazeer-ben 🙇 

I've left a few comments inline on some issues I spotted after the latest updates.

benazeer-ben committed:
Update packages/block-library/src/embed/block.json

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>

benazeer-ben committed:
Feedback changes - added wrapperProps

benazeer-ben commented:
Hi @aaronrobertshaw 

All the feedback and suggestions discussed in this PR have been addressed. I believe it is ready to go. Could you please have a final review?

Thanks in advance!

aaronrobertshaw commented:
Hi @benazeer-ben 👋 

Unfortunately, I don't have the bandwidth for reviews and testing in the near future. 

You may need to ping a few other reviewers to get this one across the line. Perhaps @t-hamano might be able to help although I know Aki is quite busy as well. Other options could be to see who else has been active around the [design tool consistency](https://github.com/WordPress/gutenberg/issues/43241) efforts and ask for them to review.

t-hamano commented:
Thanks for the ping!

At first glance, it seems like there are plenty of discussions in this PR, so I need to fully understand them before I can provide additional feedback or test them.

For now, I might not have enough time, so I can't promise a specific date, but I'd like to add it to my to-do list.
